File: internal/adapter/payment/mercadopago.go
package payment

import (
	"context"
	"fmt"

	mpconfig "github.com/mercadopago/sdk-go/pkg/config"
	mppayment "github.com/mercadopago/sdk-go/pkg/payment"
	mppreference "github.com/mercadopago/sdk-go/pkg/preference"

	"foto-app/internal/domain/port"
)

// MercadoPago implements port.PaymentGateway using the official MP Go SDK.
type MercadoPago struct {
	prefClient mppreference.Client
	payClient  mppayment.Client
	sandbox    bool
}

// NewMercadoPago creates a MercadoPago adapter using the MP SDK.
func NewMercadoPago(accessToken string, sandbox bool) (*MercadoPago, error) {
	cfg, err := mpconfig.New(accessToken)
	if err != nil {
		return nil, fmt.Errorf("create mp config: %w", err)
	}

	return &MercadoPago{
		prefClient: mppreference.NewClient(cfg),
		payClient:  mppayment.NewClient(cfg),
		sandbox:    sandbox,
	}, nil
}

// CreatePreference creates a Checkout Pro preference for the given items.
func (mp *MercadoPago) CreatePreference(
	ctx context.Context,
	externalRef string,
	items []port.MPPreferenceItem,
	backURLs *port.MPBackURLs,
) (*port.MPPreference, error) {
	sdkItems := make([]mppreference.ItemRequest, len(items))
	for i, it := range items {
		unitPrice, _ := it.UnitPrice.Float64()
		sdkItems[i] = mppreference.ItemRequest{
			Title:      it.Title,
			Quantity:   it.Quantity,
			CurrencyID: it.CurrencyID,
			UnitPrice:  unitPrice,
		}
	}

	req := mppreference.Request{
		Items:             sdkItems,
		ExternalReference: externalRef,
		BinaryMode:        true,
	}

	if backURLs != nil {
		req.BackURLs = &mppreference.BackURLsRequest{
			Success: backURLs.Success,
			Pending: backURLs.Pending,
			Failure: backURLs.Failure,
		}
	}

	resp, err := mp.prefClient.Create(ctx, req)
	if err != nil {
		return nil, fmt.Errorf("mp create preference: %w", err)
	}

	initPoint := resp.InitPoint
	if mp.sandbox && resp.SandboxInitPoint != "" {
		initPoint = resp.SandboxInitPoint
	}

	return &port.MPPreference{
		ID:        resp.ID,
		InitPoint: initPoint,
	}, nil
}

// GetPaymentInfo retrieves payment details from MP by payment ID.
func (mp *MercadoPago) GetPaymentInfo(ctx context.Context, paymentID int) (*port.MPPaymentInfo, error) {
	resp, err := mp.payClient.Get(ctx, paymentID)
	if err != nil {
		return nil, fmt.Errorf("mp get payment %d: %w", paymentID, err)
	}

	return &port.MPPaymentInfo{
		ID:                resp.ID,
		Status:            resp.Status,
		StatusDetail:      resp.StatusDetail,
		ExternalReference: resp.ExternalReference,
		TransactionAmount: resp.TransactionAmount,
	}, nil
}

File: backend/src/lib/mercadopago.js
import { MercadoPagoConfig, Preference, Payment } from 'mercadopago';
import crypto from 'crypto';

export class MercadoPagoGateway {
  constructor(accessToken, sandbox = true, webhookSecret = '') {
    this.client = new MercadoPagoConfig({
      accessToken,
      options: { timeout: 10000 },
    });
    this.preference = new Preference(this.client);
    this.payment = new Payment(this.client);
    this.sandbox = sandbox;
    this.webhookSecret = webhookSecret;
  }

  /**
   * Validate Mercado Pago webhook X-Signature header.
   * Uses HMAC-SHA256 with the webhook secret (configured in MP dashboard).
   * Reference: https://www.mercadopago.com.ar/developers/en/docs/your-integrations/notifications/webhooks
   */
  validateWebhookSignature(headers, rawBody) {
    const xSignature = headers['x-signature'];
    const xRequestId = headers['x-request-id'];
    if (!xSignature || !xRequestId) {
      // If no webhook secret is configured, skip validation (dev mode)
      if (!this.webhookSecret) return true;
      return false;
    }

    // Parse X-Signature: ts=...,v1=...
    const parts = {};
    xSignature.split(',').forEach(p => {
      const [key, val] = p.trim().split('=');
      if (key && val) parts[key.trim()] = val.trim();
    });

    const ts = parts['ts'];
    const hash = parts['v1'];
    if (!ts || !hash) return false;

    // Build manifest: "id:{data.id};request-id:{x-request-id};ts:{ts};"
    let dataId;
    try {
      const body = JSON.parse(rawBody);
      dataId = body.data?.id;
    } catch { return false; }
    if (!dataId) return false;

    const manifest = `id:${dataId};request-id:${xRequestId};ts:${ts};`;

    const expectedHash = crypto
      .createHmac('sha256', this.webhookSecret)
      .update(manifest)
      .digest('hex')
      .toLowerCase();

    return hash.toLowerCase() === expectedHash;
  }

  async createPreference(externalRef, items, backURLs = {}) {
    const mpItems = items.map((item) => ({
      title: item.title,
      quantity: item.quantity,
      unit_price: Number(item.unitPrice),
      currency_id: item.currencyId || 'ARS',
    }));

    const body = {
      items: mpItems,
      external_reference: externalRef,
      binary_mode: true,
      back_urls: {
        success: backURLs.success || '',
        pending: backURLs.pending || '',
        failure: backURLs.failure || '',
      },
    };

    const result = await this.preference.create({ body });
    const initPoint = this.sandbox && result.sandbox_init_point
      ? result.sandbox_init_point
      : result.init_point;

    return {
      id: result.id,
      initPoint,
    };
  }

  async getPaymentInfo(paymentId) {
    const result = await this.payment.get({ id: String(paymentId) });
    return {
      id: parseInt(result.id, 10),
      status: result.status,
      statusDetail: result.status_detail,
      externalReference: result.external_reference,
      transactionAmount: result.transaction_amount,
    };
  }
}

/**
 * Singleton gateway — initialized once by server.js, importable from anywhere.
 */
let _gateway = null;

export function initGateway(accessToken, sandbox, webhookSecret) {
  _gateway = new MercadoPagoGateway(accessToken, sandbox, webhookSecret);
  return _gateway;
}

export function getGateway() {
  if (!_gateway) throw new Error('MercadoPagoGateway not initialized — call initGateway first');
  return _gateway;
}
